index.js

```diff
@@ -1,135 +1,170 @@
+  return Buffer.from(
+    pubkeys.map((pub) => fromBuffer(mcl.G1, pub))
+    .reduce((acc, val) => mcl.add(acc, val))
+    .serialize()
+  )
 }
 
-// s:mcl.Fr
-// msg:string
-// return sig:mcl.G2
-const sign = (s, msg, domain) => {
-	let str = `${msg}${domain}`
-	let hash = mcl.hashAndMapToG2(str)
-	return mcl.mul(hash, s)
+/**
+ * @param {Array<bytes96>} signatures
+ * @returns {bytes96}
+ */
+function aggregateSignatures(signatures) {
+  return Buffer.from(
+    signatures.map((sig) => fromBuffer(mcl.G2, sig))
+    .reduce((acc, val) => mcl.add(acc, val))
+    .serialize()
+  )
 }
 
-// list can be either pubkeys:mcl.G1[] or signatures:mcl.G2[]
-// if pubkeys, returns aggregated pubkey P:mcl.G1
-// if signatures, returns aggregated signature Sig:mc;.G2
-const aggregate = (list) => {
-	let res = list[0]
-	for (let i = 1; i < list.length; i++) {
-		res = mcl.add(res, list[i])
-	}
-	return res
+/**
+ * @param {bytes48} pubkey
+ * @param {bytes32} messageHash
+ * @param {bytes96} signature
+ * @param {int} domain
+ * @returns {boolean}
+ */
+function verify (pubkey, messageHash, signature, domain) {
+  const pubkeyG1 = fromBuffer(mcl.G1, pubkey)
+  const signatureG2 = fromBuffer(mcl.G2, signature)
+	return mcl.pairing(pubkeyG1, hashToG2(messageHash, domain)).isEqual(mcl.pairing(g1(), signatureG2))
 }
 
-// pubkey:bls.PublicKey()
-// message:string
-// signature:bls.Signature()
-// domain:uint
-const verify = (pub, message, sig, domain) => {
-	if (!(pub instanceof mcl.G1)) throw new Error('verify:wrong pubkey type')
-	if (!(sig instanceof mcl.G2)) throw new Error('verify:wrong signature type')
-	let msg = `${message}${domain}`
-	let hash = mcl.hashAndMapToG2(msg)
-	return mcl.pairing(pub, hash).isEqual(mcl.pairing(Q(), sig))
+/**
+ * @param {Array<bytes48>} pubkeys
+ * @param {Array<bytes32>} messageHashes
+ * @param {bytes96} signature
+ * @param {int} domain
+ * @returns {boolean}
+ */
+function verifyMultiple (pubkeys, messageHashes, signature, domain) {
+	const pubkeyG1s = pubkeys.map((pub) => fromBuffer(mcl.G1, pub))
+  const signatureG2 = fromBuffer(mcl.G2, signature)
+	if (pubkeys.length !== messageHashes.length) {
+    throw new Error('number of pubkeys does not equal number of message hashes')
+  }
+
+  let ePH = pairing(pubkeyG1s[0], messageHashes[0], domain)
+  for (let i = 1; i < messageHashes.length; i++) {
+    ePH = mcl.mul(ePH, pairing(pubkeyG1s[i], messageHashes[i], domain))
+  }
+	return ePH.isEqual(mcl.pairing(g1(), signatureG2))
 }
 
-// pubkeys:list[mcl.G1]
-// messages:list[string]
-// signature:mcl.G2
-// domain:int
-const verify_multiple = (pubkeys, messages, signature, domain) => {
-	pubkeys.map((pub) => {
-        	if (!(pub instanceof mcl.G1)) throw new Error('verify_multiple:wrong pubkey type')
-	})
-	if (!(signature instanceof mcl.G2)) throw new Error('verify_multiple:wrong signature type')
-	if (pubkeys.length !== messages.length) throw new Error('verify_multiple:number of pubkeys does not equal number of messages')
```

Comment:
same with these

Reply:
see above comment, I'm thinking we can do a similar thing by checking the length of the input Buffers.